crates/subspace-farmer/src/archiving.rs

```diff
@@ -94,27 +97,50 @@ impl Archiving {
-        tokio::task::spawn_blocking({
-            move || {
-                let mut last_archived_segment_index = None;
-                while let Ok((archived_segment, acknowledgement_sender)) =
-                    archived_segments_sync_receiver.recv()
-                {
-                    let segment_index = archived_segment.root_block.segment_index();
-                    if last_archived_segment_index == Some(segment_index) {
-                        continue;
+        if let Some(node) = node {
+            tracing::trace!("Subscribing to pubsub archiving...");
+            let mut subscription = node
+                .subscribe(subspace_networking::PUB_SUB_ARCHIVING_TOPIC.clone())
+                .await?;
+            // TODO: Make sure this task can be cancelled
+            tokio::spawn(async move {
+                while let Some(bytes) = subscription.next().await {
+                    match Decode::decode(&mut bytes.as_ref()) {
+                        Ok(archived_segment) => {
+                            if plot_segment(archived_segment) == ControlFlow::Break(()) {
+                                break;
+                            }
+                        }
+                        Err(error) => tracing::error!(%error, "Failed to decode archived segment"),
                     }
-                    last_archived_segment_index.replace(segment_index);
+                    info!("Archiving segment received from pubsub subscription.");
+                }
+            });
+        } else {
+            // Erasure coding in archiver and piece encoding are CPU-intensive operations.
+            tokio::task::spawn_blocking({
+                move || {
+                    let mut last_archived_segment_index = None;
+                    while let Ok((archived_segment, acknowledgement_sender)) =
+                        archived_segments_sync_receiver.recv()
+                    {
+                        let segment_index = archived_segment.root_block.segment_index();
+                        if last_archived_segment_index == Some(segment_index) {
+                            continue;
+                        }
+                        last_archived_segment_index.replace(segment_index);
 
-                    plot_segment(archived_segment);
+                        if plot_segment(archived_segment) == ControlFlow::Break(()) {
+                            break;
+                        }
 
-                    if let Err(()) = acknowledgement_sender.send(()) {
-                        error!("Failed to send archived segment acknowledgement");
+                        if let Err(()) = acknowledgement_sender.send(()) {
+                            error!("Failed to send archived segment acknowledgement");
+                        }
                     }
                 }
-            }
-        });
+            });
+        }
 
         info!("Subscribing to archived segments");
         let mut archived_segments = client
```

Comment:
Since archived segments subscription still happens unconditionally it should produce a lot of errors due to `archived_segments_sync_sender` trying to send data over the channel whose receiver is already dropped.